crates/accelerate/src/high_level_synthesis.rs

```diff
@@ -171,82 +213,807 @@ impl QubitTracker {
-    fn to_local_mapping(&self) -> HashMap<usize, usize> {
-        HashMap::from_iter(
-            self.local_to_global
-                .iter()
-                .enumerate()
-                .map(|(i, j)| (*j, i)),
+    fn get_hls_op_names(&self) -> Vec<String> {
+        self.hls_op_names.clone()
+    }
+
+    fn get_coupling_map(&self) -> &Py<PyAny> {
+        &self.coupling_map
+    }
+
+    fn get_target(&self, py: Python) -> Option<Py<Target>> {
+        self.target.as_ref().map(|target| target.clone_ref(py))
+    }
+
+    fn get_equivalence_library(&self, py: Python) -> Option<Py<EquivalenceLibrary>> {
+        self.equivalence_library
+            .as_ref()
+            .map(|eqlib| eqlib.clone_ref(py))
+    }
+
+    fn get_device_insts(&self) -> HashSet<String> {
+        self.device_insts.clone()
+    }
+
+    fn get_use_qubit_indices(&self) -> bool {
+        self.use_qubit_indices
+    }
+
+    fn get_min_qubits(&self) -> usize {
+        self.min_qubits
+    }
+
+    fn get_unroll_definitions(&self) -> bool {
+        self.unroll_definitions
+    }
+
+    fn set_device_insts(&mut self, device_insts: HashSet<String>) {
+        self.device_insts = device_insts;
+    }
+
+    fn __str__(&self) -> String {
+        format!(
+            "HighLevelSynthesisData(hls_config: {:?}, hls_plugin_manager: {:?}, hls_op_names: {:?}, coupling_map: {:?}, target: {:?},  equivalence_library: {:?}, device_insts: {:?}, use_qubit_indices: {:?}, min_qubits: {:?}, unroll_definitions: {:?})",
+            self.hls_config, self.hls_plugin_manager, self.hls_op_names, self.coupling_map, self.target, self.equivalence_library, self.device_insts,  self.use_qubit_indices, self.min_qubits, self.unroll_definitions
         )
     }
+}
+
+/// Check whether an operation is natively supported.
+fn instruction_supported(
+    py: Python,
+    data: &Bound<HighLevelSynthesisData>,
+    name: &str,
+    qubits: &[Qubit],
+) -> bool {
+    match &data.borrow().target {
+        Some(target) => {
+            let target = target.borrow(py);
+            if target.num_qubits.is_some() {
+                if data.borrow().use_qubit_indices {
+                    let physical_qubits = qubits
+                        .iter()
+                        .map(|q| PhysicalQubit(q.index() as u32))
+                        .collect();
+                    target.instruction_supported(name, Some(&physical_qubits))
+                } else {
+                    target.instruction_supported(name, None)
+                }
+            } else {
+                data.borrow().device_insts.contains(name)
+            }
+        }
+        None => data.borrow().device_insts.contains(name),
+    }
+}
+
+/// Check whether an operation does not need to be synthesized.
+fn definitely_skip_op(
+    py: Python,
+    data: &Bound<HighLevelSynthesisData>,
+    op: &PackedOperation,
+    qubits: &[Qubit],
+) -> bool {
+    if qubits.len() < data.borrow().min_qubits {
```

Comment:
If you believe you will call the `borrow()` method more than once, make sure to just do it once and store the reference in a variable.
```suggestion
    let borrowed_data = data.borrow();
    if qubits.len() < borrowed_data.min_qubits {
```